components/webext-storage/src/db.rs

```diff
@@ -62,6 +62,37 @@ impl StorageDb {
             }
         }
     }
+
+    /// Closes the underlying database connection. If this fails, the connection
+    /// will be leaked until shutdown.
+    pub fn teardown(self) -> Result<()> {
+        Ok(self.writer.close().map_err(|(writer, err)| {
+            // Failing to close the connection means we have unfinalized
+            // statements. Since we're likely being called as the app is
+            // shutting down, anyway, we leak the connection here, and rely on
+            // the OS to clean it up. Note that the connection's destructor will
+            // try to close it again, and panic if it can't. This is fine for
+            // the FFI, but not Desktop, which is compiled with
+            // `panic = "abort"`. To avoid a shutdown crash, we use `forget`
+            // instead of `drop`.
+            mem::forget(writer);
```

Comment:
@thomcc, I'd love your feedback on this especially! ISTM a leak at shutdown (when resources will get reclaimed momentarily, anyway) is much better than a crash—are there consequences to forgetting a `rusqlite::Connection` like this, though? The FFI won't use this method, it'll go through the destructor, so that should panic if closing fails...but your `ConcurrentHandleMap` handles that perfectly.

I also thought about putting the `forget` in Desktop...but then I'd have to change the signature to return `Result<(), (StorageDb, Error)>`, and bubble that up in `Store`...and it feels like a lot of clunky code for (what we hope 🤞) is an edge case.

Reply:
Actually, I changed my mind about this. There's no guarantee that `teardown` will be called—if Desktop drops its `StorageSyncArea` (which is holding on to a `LazyStore`, which holds on to a `webext_storage::Store`), the connection will get closed via `InnerConnection::drop`, not `teardown`, causing a panic. Let's change this to use `mem::drop` instead and make it less surprising; we can always change back if we see shutdown crashes.

Reply:
But then, if we rely on `InnerConnection::drop` to close the connection...we don't really need an explicit `teardown` or `close` method, since dropping our last reference to the store should be enough. So we can remove the whole second commit and close #3014 as "won't fix". ✨

Reply:
Isn't it an xpcom component holding the reference? If so, "dropping the last reference" can often be more difficult than you think it should be :)